Xamarin.Essentials/Permissions/Permissions.ios.tvos.macos.cs

```diff
@@ -0,0 +1,101 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Photos;
+
+namespace Xamarin.Essentials
+{
+    public static partial class Permissions
+    {
+        public partial class Photos : BasePlatformPermission
+        {
+            protected override Func<IEnumerable<string>> RequiredInfoPlistKeys =>
+                () => new string[] { "NSPhotoLibraryUsageDescription" };
+
+            public override Task<PermissionStatus> CheckStatusAsync()
+            {
+                EnsureDeclared();
+
+                return Task.FromResult(GetPhotoPermissionStatus(PHAccessLevel.ReadWrite));
+            }
+
+            public override async Task<PermissionStatus> RequestAsync()
+            {
+                EnsureDeclared();
+
+                var status = GetPhotoPermissionStatus(PHAccessLevel.ReadWrite);
+                if (status == PermissionStatus.Granted)
+                {
+                    return status;
+                }
+#if __IOS__
+                else if (status == PermissionStatus.Limited)
+                {
+                    PhotosUI.PHPhotoLibrary_PhotosUISupport.PresentLimitedLibraryPicker(
+                        PHPhotoLibrary.SharedPhotoLibrary,
+                        Platform.GetCurrentUIViewController());
```

Comment:
This actually pops up a dialog that requires user input. If we return without waiting, then potentially the app will try access photos that the user is still selecting/deselecting.

There is an overload that takes a completion / awaitable:

```cs
await PHPhotoLibrary.SharedPhotoLibrary.PresentLimitedLibraryPickerAsync(Platform.GetCurrentUIViewController());
```

Reply:
hmm, I don't know. Maybe I'm `await` your review😂

Reply:
I think I didn't notice that method when I was doing it. But now I get an exception when using the async overload and a repeated request to select photos:

[crash1.log](https://github.com/xamarin/Essentials/files/7918061/crash1.log)
[crash2.log](https://github.com/xamarin/Essentials/files/7918063/crash2.log)

>System.InvalidOperationException: 
  at System.Threading.Tasks.TaskCompletionSource`1[TResult].SetResult (TResult result) [0x00009] in <dfdfd170633c4311845cda2ff4af9f6b>:0
  at PhotosUI.PHPhotoLibrary_PhotosUISupport+<>c__DisplayClass3_0.<PresentLimitedLibraryPickerAsync>b__0 (System.String[] obj_) [0x00000] in /Users/builder/azdo/_work/1/s/xamarin-macios/src/build/ios/native/PhotosUI/PHPhotoLibrary_PhotosUISupport.g.cs:104
  at ObjCRuntime.Trampolines+SDActionArity1V5.Invoke (System.IntPtr block, System.IntPtr obj) [0x00016] in /Users/builder/azdo/_work/1/s/xamarin-macios/src/build/ios/native/ObjCRuntime/Trampolines.g.cs:21418
  at at (wrapper native-to-managed) ObjCRuntime.Trampolines+SDActionArity1V5.Invoke(intptr,intptr)
  at at (wrapper managed-to-native) UIKit.UIApplication.UIApplicationMain(int,string[],intptr,intptr)
  at UIKit.UIApplication.Main (System.String[] args, System.Type principalClass, System.Type delegateClass) [0x0003b] in /Users/builder/azdo/_work/1/s/xamarin-macios/src/UIKit/UIApplication.cs:83
  at Samples.iOS.Application.Main (System.String[] args) [0x00001] in /Users/dmitriidimov/Documents/projects/Essentials/Samples/Samples.iOS/Main.cs:9

https://user-images.githubusercontent.com/59065470/150633976-cebd41c6-3eb0-4892-90d5-b996b6aceafe.MP4

Reply:
The stack trace is `SetResult` which will throw if the task is already completed. This may be a bug or something in iOS, so maybe open an issue there and then we can see what they say. The exception is in the generated code and maybe something is going wrong in the await part.